ws/mod.ts

```diff
@@ -403,79 +402,86 @@ export function createSecKey(): string {
-      throw new Error("currently https/wss is not supported");
-    }
-  }
-  const conn = await Deno.dial("tcp", `${hostname}:${port}`);
-  const abortHandshake = (err: Error): void => {
-    conn.close();
-    throw err;
-  };
-  const bufWriter = new BufWriter(conn);
-  const bufReader = new BufReader(conn);
-  await bufWriter.write(
-    encode(`GET ${pathname}?${searchParams || ""} HTTP/1.1\r\n`)
-  );
   const key = createSecKey();
+
   if (!headers.has("host")) {
     headers.set("host", hostname);
   }
   headers.set("upgrade", "websocket");
   headers.set("connection", "upgrade");
   headers.set("sec-websocket-key", key);
-  let headerStr = "";
+
+  let headerStr = `GET ${pathname}?${searchParams || ""} HTTP/1.1\r\n`;
   for (const [key, value] of headers) {
     headerStr += `${key}: ${value}\r\n`;
   }
   headerStr += "\r\n";
+
   await bufWriter.write(encode(headerStr));
-  let err, statusLine, responseHeaders;
-  err = await bufWriter.flush();
-  if (err) {
-    throw new Error("ws: failed to send handshake: " + err);
-  }
+  await bufWriter.flush();
+
   const tpReader = new TextProtoReader(bufReader);
-  [statusLine, err] = await tpReader.readLine();
-  if (err) {
-    abortHandshake(new Error("ws: failed to read status line: " + err));
+  const statusLine = await tpReader.readLine();
+  if (statusLine === EOF) {
+    throw new UnexpectedEOFError();
   }
-  const m = statusLine.match(/^(.+?) (.+?) (.+?)$/);
+  const m = statusLine.match(/^(?<version>\S+) (?<statusCode>\S+) /);
   if (!m) {
-    abortHandshake(new Error("ws: invalid status line: " + statusLine));
+    throw new Error("ws: invalid status line: " + statusLine);
   }
-  const [_, version, statusCode] = m;
+  const { version, statusCode } = m.groups;
   if (version !== "HTTP/1.1" || statusCode !== "101") {
-    abortHandshake(
-      new Error(
-        `ws: server didn't accept handshake: version=${version}, statusCode=${statusCode}`
-      )
+    throw new Error(
+      `ws: server didn't accept handshake: ` +
+        `version=${version}, statusCode=${statusCode}`
     );
   }
-  [responseHeaders, err] = await tpReader.readMIMEHeader();
-  if (err) {
-    abortHandshake(new Error("ws: failed to parse response headers: " + err));
+
+  const responseHeaders = await tpReader.readMIMEHeader();
+  if (responseHeaders === EOF) {
+    throw new UnexpectedEOFError();
   }
+
   const expectedSecAccept = createSecAccept(key);
   const secAccept = responseHeaders.get("sec-websocket-accept");
   if (secAccept !== expectedSecAccept) {
-    abortHandshake(
-      new Error(
-        `ws: unexpected sec-websocket-accept header: expected=${expectedSecAccept}, actual=${secAccept}`
-      )
+    throw new Error(
+      `ws: unexpected sec-websocket-accept header: ` +
+        `expected=${expectedSecAccept}, actual=${secAccept}`
     );
   }
+}
+
+/** Connect to given websocket endpoint url. Endpoint must be acceptable for URL */
+export async function connectWebSocket(
```

Comment:
It seems like you switched the order of handshake and connectWebSocket .. which is making for a difficult to read diff. Consider switching order back.

Reply:
The `handshake()` function is new. I've moved it after `connectWebSocket()`, this does indeed make the diff a bit more readable.

Reply:
Had to switch the order back because of tslint's no-use-before-define rule.